### ISSUE TITLE
TransportGetAllocationStatsAction may cause significant load on elected master

### ISSUE THREAD

**Issue by idegtiarenko** [elastic/elasticsearch#110716]
### Elasticsearch Version

8.14

### Java Version

_bundled_

### OS Version

any

### Problem Description

TransportGetAllocationStatsAction runs on elected master so it is theoretically possible to overload it by executing node stats requests around various nodes in cluster, especially in a clusters with many shards as complexity is proportional to the shard count.

The aggregated result of the computation is small (5 numbers per node), we should consider caching it for small period of time (1 minute?) and reuse it between different calls during.

https://github.com/elastic/elasticsearch/blob/5e5205994747618edf5d81cb6a813adef9f2dbce/server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/TransportNodesStatsAction.java#L88-L96

### Steps to Reproduce

n/a

**Comment by mhl-b** [comment]
Do we need client caching too? Server caching should help when there are few nodes with large number of shards. But if there many nodes that frequently poll stats, networking overhead will cripple up. May be distribute load, client picks random node from cluster and if that node does not have cache, forward to master and populate cache.

**Comment by DaveCTurner** [comment]
`TransportGetAllocationStatsAction$Response` is pretty small, ~100B/node or thereabouts, I think caching it on the master is sufficient.

**Comment by shreedaddy** [comment]
> ### Elasticsearch Version
> 8.14
> 
> ### Installed Plugins
> _No response_
> 
> ### Java Version
> _bundled_
> 
> ### OS Version
> any
> 
> ### Problem Description
> TransportGetAllocationStatsAction runs on elected master so it is theoretically possible to overload it by executing node stats requests around various nodes in cluster, especially in a clusters with many shards as complexity is proportional to the shard count.
> 
> The aggregated result of the computation is small (5 numbers per node), we should consider caching it for small period of time (1 minute?) and reuse it between different calls during.
> 
> https://github.com/elastic/elasticsearch/blob/5e5205994747618edf5d81cb6a813adef9f2dbce/server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/TransportNodesStatsAction.java#L88-L96
> 
> ### Steps to Reproduce
> n/a
> 
> ### Logs (if relevant)
> _No response_

Please let me know if this issue is open. If it is can you please assign it to me.

**Comment by shreedaddy** [comment]
> `TransportGetAllocationStatsAction$Response` is pretty small, ~100B/node or thereabouts, I think caching it on the master is sufficient.

Please let me know if this issue is open. If it is can you please assign it to me.

**Comment by DaveCTurner** [comment]
Still available @shreedaddy thanks for the offer. We can't assign issues to folks outside the @elastic org but if you want to contribute a PR then please feel free.

**Comment by shreedaddy** [comment]
> Still available @shreedaddy thanks for the offer. We can't assign issues to folks outside the @elastic org but if you want to contribute a PR then please feel free.

Will get started.

**Comment by shreedaddy** [comment]
> Still available @shreedaddy thanks for the offer. We can't assign issues to folks outside the @elastic org but if you want to contribute a PR then please feel free.

Dave I have already completed analysis but to get started on the development I do not have access to create a new branch for this issue. Can you please provide this particular privilege.

**Comment by DaveCTurner** [comment]
Hi @shreedaddy and thanks for your interest here. A PR would be welcome, but I don't really have the capacity to offer you as much help as you are asking for here. If you cannot work this out yourself from the code and surrounding documentation, please pick a different issue to address.

**Comment by shreedaddy** [comment]
> Dave I have already completed analysis but to get started on the development I do not have access to create a new branch for this issue. Can you please provide this particular privilege.

Dave I have already completed analysis but to get started on the development I do not have access to create a new branch for this issue. Can you please provide this particular privilege.

**Comment by DaveCTurner** [comment]
> Dave I have already completed analysis but to get started on the development I do not have access to create a new branch for this issue. Can you please provide this particular privilege.

No, you already have all the privileges you need to open a PR based on a branch in your own personal fork.

**Comment by dragan509** [comment]
It would be better to use TransportClusterAllocationExplainAction instead of TransportGetAllocationStatsAction.
Of course, the result must be different from what you want, but if the issue occurs, TransportClusterAllocationExplainAction is more suitable than TransportGetAllocationStatsAction.
I’m still gaining experience, so I wonder if this idea could be beneficial.
I hope you don't mind me suggesting this, as I'm still learning. I appreciate any feedback you have.
Thanks.
